### DISCUSSION TITLE
jdbc获取元数据类型问题

### DISCUSSION THREAD

**Discussion by kaiwa8** [apache/doris#24809] [A - General / Q&A] [answered]
jdbc获取表元数据，为什么varchar会被识别成char，长度也不对。char字符长度不对，largeint识别不出来

![屏幕截图 2023-09-22 203627](https://github.com/apache/doris/assets/41932093/8d16e0e4-34e6-4a59-8e35-b56d4346e815)

jdbc查询结果

String sql = "select * from test_db";
        Statement statement = connection.createStatement();
        //执行sql语句
        ResultSet rs = statement.executeQuery(sql);
        //获取ResultSetMetaData对象
        ResultSetMetaData metaData = rs.getMetaData();
        //获取查询字段数量
        int columnCount = metaData.getColumnCount();
        for (int i = 1; i <= columnCount; i++) {
            //获取表名称
            String columnName = metaData.getColumnName(i);
            //获取sql类型
            String columnTypeName = metaData.getColumnTypeName(i);
            int precision = metaData.getPrecision(i);
            System.out.println("columnName:"+columnName+" columnTypeName: "+columnTypeName+" : "+"precision:"+precision);
        }

输出

columnName:id columnTypeName: INT : precision:10
columnName:varchar2 columnTypeName: CHAR : precision:85
columnName:char2 columnTypeName: CHAR : precision:85
columnName:largeint2 columnTypeName: CHAR : precision:85

为什么varchar2字段类型是 CHAR ，precision是85？
为什么char2 precision是85？
为什么largeint2 识别出错？

**Comment by yagagagaga** [comment]
What is the version of your MySQL JDBC driver? You should try to use updated drivers as much as possible.

**Reply by kaiwa8** [reply]
8.0.28

**Reply by yagagagaga** [reply]
因为mysql没有`largeint`类型，所以Doris返回`char`类型代替，`char`和`varchar`在doris里都认为是`string`类型，其返回码是254，这在mysql jdbc driver里对应的是`char`，因此，你看到返回值是`char`。至于 precision ，仅有 decimal 类型会返回正确的值。

**Reply by kaiwa8** [reply]
好的，谢谢

**Reply by LuckyCaptain-go** [reply]
有什么兼容计划，或者解决办法吗？

**Comment by LemonLiTree** [comment, marked as answer]
https://doris.apache.org/docs/dev/admin-manual/http-actions/fe/meta-action/
It is recommended to use doris' API to obtain metadata

**Reply by kaiwa8** [reply]
好的，谢谢
